src/enveloping/EnvelopingArbiter.ts

```diff
@@ -0,0 +1,148 @@
+import { Bytes } from 'ethers'
+import log from 'loglevel'
+import { HttpProvider, IpcProvider, WebsocketProvider } from 'web3-core'
+import { Address } from '../relayclient/types/Aliases'
+import { ServerConfigParams } from '../relayserver/ServerConfigParams'
+import { TransactionManager } from '../relayserver/TransactionManager'
+import { CommitmentValidator } from './CommitmentValidator'
+import { CommitmentReceipt } from './Commitment'
+import { FeeEstimator, FeesTable } from './FeeEstimator'
+import { NonceQueueSelector } from './NonceQueueSelector'
+
+export type Web3Provider =
+  | HttpProvider
+  | IpcProvider
+  | WebsocketProvider
+
+/**
+ * The main Enveloping Arbiter component. It is in charge of ensuring quality of service
+ * for the Enveloping clients. It adds multiple workers support through the NonceQueueSelector, each
+ * corresponding to a different "tier" of service. It also provides assurance to the client in the form
+ * of a signed Commitment receipt, that can be validated through the CommitmentValidator or a
+ * smart contract in the future. A Fee Estimator monitors the blockchain and estimates the recommended
+ * gas prices for different tolerated delays.
+ */
+export class EnvelopingArbiter {
+  readonly commitmentValidator: CommitmentValidator
+  readonly feeEstimator: FeeEstimator
+  readonly nonceQueueSelector: NonceQueueSelector
+
+  constructor (config: Partial<ServerConfigParams>, provider: Web3Provider) {
+    this.commitmentValidator = new CommitmentValidator()
+    this.feeEstimator = new FeeEstimator(config, provider)
+    this.nonceQueueSelector = new NonceQueueSelector()
+  }
+
+  /**
+   * Checks if the given timestamp is valid. If not, it defaults to 5min in the future
+   * @param maxTime the timestamp to validate
+   * @return a valid timestamp
+   */
+  checkMaxDelayForResponse (maxTime?: string): number {
+    if (typeof maxTime === 'undefined' || !this.isValidTime(maxTime)) {
+      maxTime = (Date.now() + (300 * 1000)).toString()
+    }
+    return parseInt(maxTime)
+  }
+
+  /**
+   * Waits the given number of miliseconds
+   * @param ms the time to wait
+   */
+  async delay (ms: number): Promise<void> {
+    return await new Promise((resolve) => {
+      setTimeout(() => resolve(), ms)
+    })
+  }
+
+  /**
+   * Checks if the given timestamp is valid and not from the past
+   */
+  isValidTime (maxTime: string): boolean {
+    return parseInt(maxTime) > Date.now()
+  }
+
+  /**
+   * Reads the Fees Table from the Fee Estimator and returns it. When first started, the
+   * Fee Estimator will take a few seconds to process the data, meanwhile the fees table
+   * won't be available. In this case, it will keep checking every second until it is.
+   */
+  async getFeesTable (): Promise<FeesTable> {
+    if (typeof this.feeEstimator.feesTable !== 'undefined') {
+      return this.feeEstimator.feesTable
+    } else {
+      await this.delay(1000)
+      return await this.getFeesTable()
+    }
+  }
```

Comment:
let's look for an alternative here. this is potentially dangerous

Reply:
We should put a TODO maybe but these changes are coming from the original PR.